src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs

```diff
@@ -0,0 +1,264 @@
+            // the user from squiggles or error-list.
+            HandleRemovedDocuments(previousResults, progress);
+
+            // Create a mapping from documents to the previous results the client says it has for them.  That way as we
+            // process documents we know if we should tell the client it should stay the same, or we can tell it what
+            // the updated diagnostics are.
+            var documentToPreviousDiagnosticParams =
+                previousResults.Select(r => new { DiagnosticParams = r, Document = _solutionProvider.GetDocument(r.TextDocument) })
+                               .Where(a => a.Document != null)
+                               .ToDictionary(a => a.Document, a => a.DiagnosticParams);
+
+            // Next process each file in priority order. Determine if diagnostics are changed or unchanged since the
+            // last time we notified the client.  Report back either to the client so they can update accordingly.
+            foreach (var document in GetOrderedDocuments(context))
+            {
+                if (DiagnosticsAreUnchanged(documentToPreviousDiagnosticParams, document))
+                {
+                    // Nothing changed between the last request and this one.  Report a (null-diagnostics,
+                    // same-result-id) response to the client as that means they should just preserve the current
+                    // diagnostics they have for this file.
+                    var previousParams = documentToPreviousDiagnosticParams[document];
+                    progress.Report(CreateReport(previousParams.TextDocument, diagnostics: null, previousParams.PreviousResultId));
+                }
+                else
+                {
+                    await ComputeAndReportCurrentDiagnosticsAsync(progress, document, cancellationToken).ConfigureAwait(false);
+                }
+            }
+
+            // If we had a progress object, then we will have been reporting to that.  Otherwise, take what we've been
+            // collecting and return that.
+            return progress.GetValues();
+        }
+
+        private async Task ComputeAndReportCurrentDiagnosticsAsync(
+            BufferedProgress<TReport> progress,
+            Document document,
+            CancellationToken cancellationToken)
+        {
+            // Being asked about this document for the first time.  Or being asked again and we have different
+            // diagnostics.  Compute and report the current diagnostics info for this document.
+
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            using var _ = ArrayBuilder<VSDiagnostic>.GetInstance(out var result);
+
+            var diagnostics = _diagnosticService.GetDiagnostics(document, includeSuppressedDiagnostics: false, forPullDiagnostics: true, cancellationToken);
```

Comment:
this is new code. it calls `forPullDiagnostics: true` as it wants to only run if we're explicitly opted-into pull mode.